From Linux bash script to Python control flow

joerick opened the pull request:
We've been talking for a while about moving more of the Linux control flow into Python, instead of bash. This PR explores that. 

This is a first implementation of that idea. It has a build script that resembles that of macOS or Windows, but runs every command through `docker exec`.

Code-wise, I think this is a nice improvement. It avoids lots of bash quirks and fiddling with strings. Hopefully, it should unlock lots of improvements down the line.

The downside, currently, is performance. Those `docker exec` calls are pretty slow, compared to local subprocess calls. Each `docker exec` takes around 150ms on my machine. We'll see how it shakes out in our CI providers, but in my testing it looks like roughly +33% run time on the CI suite.

To be clear, the effects will be much worse on our CI than in any user scenario, because we run hundreds of tiny builds, and users probably have a handful of more substantial ones. But the CI is already pretty tedious, so I'd like to explore options to speed it up before committing to this approach.

My current thought to improve perf is to keep a long-running shell process running in the Docker container, and relay commands to that via open file descriptors. Couple details to figure out but should be fairly transparent I hope.

Comments welcome!

Czaki commented:
What did you think about create python script which will be copied into docker container and then executed in one command?

joerick commented:
> What did you think about create python script which will be copied into docker container and then executed in one command?

I think it could work, but there's loads of details to figure out. What state/configuration and code to pass over? How is passed state serialised/deserialized? How to interpret the output of the script? Which sections of the codebase can be called from that remote script? Worst case, it'd be the same as the current bash implementation, with huge amounts of string interpolation, writing Python code inside Python strings. I also was keen to avoid the mental overhead of having to think what sections of the codebase could be called by different files.

joerick commented:
Alright, that should be the performance problems fixed. Instead of `docker exec` I'm using a remote shell running inside the Docker container, relaying commands to it and streaming back the stdout. Works pretty well! By chance, that means we can use ENTRYPOINT for that shell, so not specify `linux32` anymore.

YannickJadoul commented on cibuildwheel/bashlex_eval.py:
```diff
@@ -1,17 +1,26 @@
 import subprocess
 
-from typing import Dict, NamedTuple
+from typing import Dict, NamedTuple, Callable, Optional
```
Sorted?

YannickJadoul commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,178 @@
+        return json.loads(self.call([
+            self.UTILITY_PYTHON,
+            '-c',
+            'import sys, json, os; json.dump(os.environ.copy(), sys.stdout)'
```
`os.environ.copy()`?

joerick replied:
`os.environ` is some weird object that json can't serialise, but `copy()` returns a `dict`.

YannickJadoul replied:
Ha, thought it would look enough like a duck to not require this. No problem, though, but I thought it was maybe something remaining from an older version.

YannickJadoul commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,178 @@
+        if from_path.is_dir():
+            self.call(['mkdir', '-p', to_path])
+            subprocess.run(
+                f'tar cf - . | docker exec -i {self.container_name} tar -xC {to_path} -f -',
```
I'm wondering if something with `base64` might work. But obviously, `tar ... | base64` will result in a huuuuuge string, so we need to make sure it can be passed nicely to `base64 -d` inside the docker.

joerick replied:
Some of the reason for this is how the solution evolved. The `call` method is really geared around strings - it wouldn't be possible to run binary data through it. So I guess that's where your base64 idea might come in... or I suppose `call` could be built on top of a binary version.

But, this solution with shell pipes is nice because pipes don't buffer in memory, are easy to write, and the archiving and unarchiving are happening concurrently.

Before this PR, these copies were mostly `docker cp` calls - I doubt this performance using `exec` is very different from that... at least, I don't see a difference in practise.

YannickJadoul commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,178 @@
+        # log the command we're executing
+        print(f'    + {command}')
+
+        # Write a command to the remote shell. First we change the
+        # cwd, if that's required. Then, we use the `env` utility to run
+        # `command` inside the specified environment. We use `env` because it
+        # can cope with spaces and strange characters in the name or value.
+        # Finally, the remote shell is told to write a footer - this will show
+        # up in the output so we know when to stop reading, and will include
+        # the returncode of `command`.
```
Or maybe https://msgpack.org/index.html?

joerick replied:
Yeah, it's a little weird, I must admit. If there's some kind of RPC system that works with UNIX stdin/stdout that would be ideal. And is preinstalled on Centos :)

To be honest, my main concern with it is the binary->text conversion that's happening with these Python streams. E.g. if a program outputs a weird non-UTF8 character sequence, what happens? Probably that's another one for the unit test list :)

YannickJadoul replied:
> To be honest, my main concern with it is the binary->text conversion that's happening with these Python streams. E.g. if a program outputs a weird non-UTF8 character sequence, what happens? Probably that's another one for the unit test list :)

Wouldn't that be solved by `base64`?

joerick replied:
That's resolved now using `bytes` streams in DockerContainer.

YannickJadoul replied:
Right, I'll give it another look, then :-)

YannickJadoul commented on cibuildwheel/linux.py:
```diff
@@ -111,201 +102,133 @@ def build(options: BuildOptions) -> None:
+                    docker.call(['sh', '-c', before_all_prepared], env=env)
+
+                for config in platform_configs:
+                    dependency_constraint_flags: List[Union[str, PathLike]] = []
```
Unrelated to this PR, I think: If you put this as `else`-clause, you won't need the annotation (since the first occurrence would have the right type).

joerick replied:
The first occurrence resolves to `List[object]` unfortunately, so needs to be explicit

YannickJadoul commented on cibuildwheel/linux.py:
```diff
@@ -111,201 +102,133 @@ def build(options: BuildOptions) -> None:
+
+                    # put this config's python top of the list
+                    python_bin = config.path / 'bin'
+                    env['PATH'] = f'{str(python_bin)}:{env["PATH"]}'
```
Do we need `str()` there? I think it might just be fine like `{python_bin}`, no?

joerick replied:
👍

YannickJadoul commented on cibuildwheel/linux.py:
```diff
@@ -111,201 +102,133 @@ def build(options: BuildOptions) -> None:
+                    docker.call(['rm', '-rf', built_wheel_dir])
+                    docker.call(['mkdir', '-p', built_wheel_dir])
+
+                    docker.call(
+                        [
```
Same here, as in the other PR. Feels like very broad formatting, but maybe that's fine.

joerick replied:
👍

joerick commented:
> The implementation is quite impressive as well, but also quite heavy, somehow, and then especially the output-retrieving part of `call`. I'm somehow not entirely convinced there isn't a better solution, and it does make me wonder if it is worth it (I can't immediately see how or where, but if there are problems or errors in there, they will potentially be very obscure?).

Agree that the docker part is kinda... low-level protocolish. I wonder if a few unit tests on that piece in isolation might make us feel better about using it?

joerick commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,178 @@
+    the command is relayed to the remote shell, and the results are streamed
+    back to cibuildwheel.
+    '''
+    UTILITY_PYTHON = '/opt/python/cp38-cp38/bin/python'
```
Feels pretty arbitrary to me. As long as it exists in the Docker image I'm happy. I'd have used /usr/bin/python but it doesn't have `json`!

YannickJadoul replied:
Also fine :-)

> I'd have used /usr/bin/python but it doesn't have `json`!

That's still 2.6. Let's not ;-)

joerick commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,178 @@
+            stdout=subprocess.PIPE,
+            # make the input buffer large enough to carry a lot of environment
+            # variables. We choose 256kB.
+            bufsize=262144,
```
When I was writing it I was thinking that the entire command would have to fit inside the buffer. But now I'm looking at it, the shell in the container should greedily read from that fd until return is reached. So I guess it probably doesn't matter after all. I'll try setting it to something really low.

This could be something we unit test in fact :)

joerick replied:
Yeah, looks like it doesn't matter. But I'll check that in a unit test.

joerick commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,178 @@
+        self.bash_stdout = process.stdout
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
```
`__exit__` is still called if an exception is raised, but not if that exception occurred in `__enter__`, so this matches the previous code, except that if the `docker attach` fails, the container won't be cleaned up. Considering that we weren't too worried about leaking containers, I think that's okay.

YannickJadoul replied:
Ah, yes, indeed. Alright, then; let's not worry, then! :-)

joerick commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,178 @@
+            cwd=to_path
+        )
+
+    def glob(self, pattern: PurePath) -> List[PurePath]:
```
ahhh yeahh. I might want to change that. Perhaps `glob(path: Path, pattern: str)`?

YannickJadoul replied:
Maybe, yes. I think that would match better.
But it's also just a pedantic remark, because the current code does work. So it might not be worth the change either.

joerick replied:
fixed :)

YannickJadoul commented:
> I wonder if a few unit tests on that piece in isolation might make us feel better about using it?

Ah yes, that would probably already make quite a difference :-)
Though maybe we should not download/run a full docker image by default, but somehow make it optional to run? (something [like this](https://docs.pytest.org/en/stable/example/simple.html#control-skipping-of-tests-according-to-command-line-option) perhaps?)

joerick commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,178 @@
+        env_assignments = ' '.join(f'{shlex.quote(k)}={shlex.quote(v)}'
+                                   for k, v in env.items())
+        command = ' '.join(shlex.quote(str(a)) for a in args)
+        end_of_message = str(uuid.uuid4())
```
There's nothing stopping a child process from outputting `\0` and causing havoc. 

Normally, when doing this kind of protocol, there'd be a header that specifies the length of the following message -that makes reads really easy. But I wanted to avoid buffering the output on the container side, for memory reasons but mostly so that output still streamed back to the console - during a long build this is very reassuring, or helpful if a command stalls. This solution was inspired by SMTP [multipart boundaries](https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html) - basically a random string long enough is unique enough to delimit between parts of data. UUID is just an easy way to make a unique string.

I could add some of this context as a comment I guess?

YannickJadoul replied:
>  But I wanted to avoid buffering the output on the container side, for memory reasons but mostly so that output still streamed back to the console - during a long build this is very reassuring, or helpful if a command stalls.

Alright, yes. Good point! I hadn't considered that, yet, but this is an important feature, then! :-)

YannickJadoul replied:
We could still just define our own constant magic end-of-message UUID, no? Is there a reason to introduce randomness and regenerate a new one every time?

joerick replied:
I think there's something comforting about using a new UUID each time - it's a string that's never existed in the world before. but practically, err, the most concrete I can get with it is that if it was static and written in code, and then somehow this code file was sent down the pipe, then the terminator would appear before the end of the message. Far-fetched, but that's the kind of thing I'm thinking about. But a fresh UUID has never existed before, so that kind of thing couldn't happen.

YannickJadoul replied:
That's a bit far-fetched, but it had completely not crossed my mind! Better safe than sorry, then :-)

joerick commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,178 @@
+            # variables. We choose 256kB.
+            bufsize=262144,
+        )
+        self.process = process
```
👍

YannickJadoul commented on unit_test/conftest.py:
```diff
@@ -0,0 +1,21 @@
+
+def pytest_addoption(parser):
+    parser.addoption(
+        "--runslow", action="store_true", default=False, help="run slow tests"
```
I'm not sure we should have this as `--runslow`. Rather something like `--with-docker` or something? It's not per se about slow, but about not "polluting" the unit test environment with docker images (for which I meant the `--runslow` example from the pytest docs might be a good demonstration of how to achieve that)?

joerick commented:
Okie doke, this is ready for another review, if you have time @YannickJadoul !

I've fixed some the issues you raised above, and added unit tests for `DockerContainer`.

It got a little... complicated... when I realised that one of the preexisting tests `test_overridden_path` wasn't failing for the correct reason. So I have fixed that, it involved some changes to `bashlex_eval` to support `command1; command2`  syntax inside command substitutions (`command1 && command2` isn't supported due to a bug in bashlex https://github.com/idank/bashlex/issues/54). In the end, I switched to using BEFORE_ALL for that setup anyway, but there's now a unit test to cover `command1; command2` in bashlex.

Anyway, let me know what you think!

YannickJadoul commented on cibuildwheel/bashlex_eval.py:
```diff
@@ -65,9 +74,36 @@ def evaluate_word_node(node: bashlex.ast.node, context: NodeExecutionContext) ->
 
 
 def evaluate_command_node(node: bashlex.ast.node, context: NodeExecutionContext) -> str:
-    words = [evaluate_node(part, context=context) for part in node.parts]
```
What happened before, then? `';'` was just not supported either?

joerick replied:
Not within command substitutions, no.

YannickJadoul commented on test/test_environment.py:
```diff
@@ -54,8 +54,20 @@ def test_overridden_path(tmp_path):
+        else:
+            new_path = tmp_path / 'another_bin'
+            new_path.mkdir()
+            (new_path / 'python').touch(mode=0o777)
```
First time I'm seeing that octal literals are useful! :-D

joerick replied:
First time I'd used them in Python 3, too!

YannickJadoul commented on cibuildwheel/docker_container.py:
```diff
@@ -0,0 +1,189 @@
+        if from_path.is_dir():
+            self.call(['mkdir', '-p', to_path])
+            subprocess.run(
+                f'tar cf - . | docker exec -i {self.name} tar -xC {shell_quote(to_path)} -f -',
```
Just checking, since you were disappointed by the speed of `docker exec`. Using byte-streams now, it's still not possible to do this with `self.call`?

joerick replied:
Ooops, I just saw this after the merge. It would be possible, but pretty fiddly to implement. The easiest implementation would involve reading an entire tarball into memory within the host process before streaming it into the container. With a large codebase that might introduce memory issues, and could well be slower. A better implementation would stream between active `tar` processes on either side, but I'm guessing that would involve some pretty gnarly non-blocking I/O code to get it to work right. 

I was curious - here's a profile of the basic test project build - 

```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 19:54:25  Samples:  373
 /_//_/// /_\ / //_// / //_'/ //     Duration: 35.644    CPU time: 0.144
/   _/                      v3.1.3

Program: /Users/joerick/Projects/cibuildwheel/cibuildwheel/__main__.py --platform linux

35.642 <module>  __main__.py:1
└─ 35.599 main  __main__.py:56
   └─ 35.597 build  linux.py:78
      ├─ 25.233 call  docker_container.py:117
      ├─ 3.552 copy_into  docker_container.py:76
      │  └─ 3.541 run  subprocess.py:447
      │        [11 frames hidden]  subprocess
      ├─ 3.184 __enter__  docker_container.py:34
      │  ├─ 2.655 call  docker_container.py:117
      │  └─ 0.520 run  subprocess.py:447
      │        [11 frames hidden]  subprocess
      ├─ 1.854 __exit__  docker_container.py:68
      │  └─ 1.575 wait  subprocess.py:1014
      │        [4 frames hidden]  subprocess
      ├─ 0.682 glob  docker_container.py:106
      │  └─ 0.682 call  docker_container.py:117
      ├─ 0.536 copy_out  docker_container.py:95
      │  └─ 0.536 run  subprocess.py:447
      │        [9 frames hidden]  subprocess
      └─ 0.518 get_environment  docker_container.py:176
         └─ 0.518 call  docker_container.py:117

```

`copy_into` is 10% of the time here. Maybe not worth the hassle.

YannickJadoul replied:
Might not be indeed. It just looks out of place; that's why I'm asked.

joerick commented:
Thanks for looking it over, @YannickJadoul!
